nc2zarr/cli.py

```diff
@@ -138,3 +140,112 @@ def nc2zarr(input_paths: Tuple[str],
 
 if __name__ == '__main__':
     nc2zarr()
+
+
+@click.command(name='nc2zarr-batch')
+@click.argument('config_template_path', metavar='CONFIG_TEMPLATE_PATH')
+@click.argument('config_path_template', metavar='CONFIG_PATH_TEMPLATE')
+@click.option('--range', '-R', 'ranges',
+              metavar='KEY MIN MAX', nargs=3, multiple=True,
+              help=f'Key value range assignments. MIN and MAX must be integers. Option may be repeated.')
+@click.option('--value', '-V', 'values',
+              metavar='KEY VALUE', nargs=2, multiple=True,
+              help=f'Key value assignments. Option may be repeated.')
+@click.option('--scheduler', '-s', 'scheduler_config_path',
+              metavar='FILE',
+              help=f'Scheduler configuration file (YAML).')
+@click.option('--dry-run', '-d', 'dry_run', is_flag=True, default=None,
+              help='Open and process inputs only, omit data writing.')
+@click.option('--verbose', '-v', 'verbose', count=True,
+              help='Print more output. Use twice for even more output.')
+def nc2zarr_batch(
+        config_template_path: str,
+        config_path_template: str,
+        ranges: Tuple[Tuple[str]],
+        values: Tuple[Tuple[str]],
+        scheduler_config_path: str,
+        dry_run: bool,
+        verbose: int
+):
+    """
+    Run nc2zarr in batch mode.
+
+    Example:
+
+    \b
+        $ nc2zarr-batch \\
+            --range year 2002 2017 \\
+            --value base_dir . \\
+            --scheduler ../slurm-config.yml \\
+            seaice/config-template.yml \\
+            seaice/batch/config-${year}.yml
+
+    """
+    import datetime
+    import time
+    import os.path
+    import itertools
+    import yaml
+
+    from nc2zarr.batch import TemplateBatch
+
+    if not os.path.isfile(config_template_path):
+        raise click.exceptions.FileError(config_template_path, 'not found')
+
+    for k, _, _ in ranges:
+        ref = "${" + k + "}"
+        if ref not in config_path_template:
+            raise click.exceptions.BadArgumentUsage(f'reference "{ref}" '
+                                                    f'missing in CONFIG_PATH_TEMPLATE')
+
+    # noinspection PyTypeChecker
+    product_args = [range(int(min_value), int(max_value) + 1) for _, min_value, max_value in ranges] \
+                   + [(value,) for _, value in values]
+
+    keys = [k for k, _, _ in ranges] + [k for k, _ in values]
```

Comment:
If I understand correctly, keys shouldn't be repeated (either within or across the `--value` and `--range` parameters). If so, a `if len(keys) > len(set(keys)): raise ValueError("...")` might be a good safety precaution here.

Reply:
Later